pkg/ccl/sqlproxyccl/connector.go

```diff
@@ -158,9 +169,11 @@ func (c *connector) OpenTenantConnWithAuth(
 
 	// Perform user authentication for non-token-based auth methods. This will
 	// block until the server has authenticated the client.
-	if err := authenticate(clientConn, serverConn, throttleHook); err != nil {
+	cancelInfo, err := authenticate(clientConn, serverConn, throttleHook)
+	if err != nil {
 		return nil, true, err
 	}
+	c.CancelInfo = cancelInfo
```

Comment:
nit: The data flow for CancelInfo feels a little off to me. In some cases the connector constructs and assigns it, in other cases it modifies it. I also don't love the fact that we return some information as return values and some by modifying parameters on the constructor.

What do you think of this implementation:
1. When the CancelInfo is first constructed it contains no backend information.
2. The connector returns a backendConnection struct. The struct contains the net.Conn and the pgproto3.BackendKeyData.
3. CancelInfo has a .SetBackend(backendConnection) helper that records the backend data. The callers of connector are expected to update the cancel info.

Reply:
> In some cases the connector constructs and assigns it, in other cases it modifies it.

hm, i guess i view it as "the connector is responsible for setting the cancel info." so it assigns the cancel info, and modifies it when needed.

> some by modifying parameters on the constructor.

i'm not sure which parameters you're referring to

---

that said, i can give what you say a try.